server/src/main/java/org/apache/druid/segment/realtime/appenderator/BaseAppenderatorDriver.java

```diff
@@ -697,7 +703,7 @@ ListenableFuture<SegmentsAndCommitMetadata> publishInBackground(
                 Throwables.propagateIfPossible(e);
                 throw new RuntimeException(e);
               }
-              return segmentsAndCommitMetadata;
+              return segmentsAndCommitMetadata.withUpgradedSegments(upgradedSegments);
```

Comment:
We shouldn't have to distinguish between upgraded and other segments here. Instead of creating a new `SegmentsAndCommitMetadata` with upgraded segments, we should create a new `SegmentsAndCommitMetadata` which has a single set containing all the segments (upgraded and otherwise) which were committed and thus must be handed off.

```suggestion
return new SegmentsAndCommitMetadata(publishResult.getSegments(), metadata);
```

So, essentially, the original `segmentsAndCommitMetadata` object passed in to this method represents segments we wanted to commit and the new object would represent segments that were actually committed.

Reply:
This can be addressed later.